device/transport/amqp/src/amqp_twin_client.ts

```diff
@@ -118,16 +150,17 @@ export class AmqpTwinClient extends EventEmitter {
                   this._client.attachReceiverLink( this._endpoint, this._generateTwinLinkProperties(linkCorrelationId), (receiverLinkError?: Error, receiverTransportObject?: any): void => {
                     if (receiverLinkError) {
                       /* Codes_SRS_NODE_DEVICE_AMQP_TWIN_06_022: [If an error occurs on establishing the upstream or downstream link then the `error` event shall be emitted.] */
-                      this._fsm.handle('handleErrorEmit', receiverLinkError);
+                      this._fsm.transition('detached', receiverLinkError, attachCallback);
                     } else {
-                      this._downstreamAmqpLink = receiverTransportObject;
-                      this._downstreamAmqpLink.on('error', this._handleError.bind(this));
+                      this._receiverLink = receiverTransportObject;
+                      this._receiverLink.on('message', this._messageHandler);
+                      this._receiverLink.on('error', this._errorHandler);
                       this._client.attachSenderLink( this._endpoint, this._generateTwinLinkProperties(linkCorrelationId), (senderLinkError?: Error, senderTransportObject?: any): void => {
                         if (senderLinkError) {
-                          this._fsm.handle('handleErrorEmit', senderLinkError);
+                          this._fsm.transition('detached', senderLinkError, attachCallback);
```

Comment:
>transition [](start = 36, length = 10)

detach the receiver link if the sender link fails to attach?

Reply:
yes in theory. in practice if a link is being detached the whole thing is going down and detach is going to be handled by the lower layer. I'm not satisfied with how we carry links around in the AMQP stack for now but this would be the subject of a bigger refactoring. 

---
In reply to: [171687889](https://github.com/pierreca/azure-iot-sdk-node/pull/1#discussion_r171687889) [](ancestors = 171687889)